udev_monitor.c

```diff
@@ -116,25 +110,21 @@ static void *udev_monitor_handle_event(void *ptr)
 {
     struct udev_monitor *udev_monitor = ptr;
     struct inotify_event *event;
-    struct epoll_event epoll[2];
+    struct pollfd poll_fd;
     char data[4096];
-    sigset_t set;
     ssize_t len;
     int i;
 
-    sigfillset(&set);
-    pthread_sigmask(SIG_BLOCK, &set, NULL);
-
-    while (epoll_wait(udev_monitor->efd, epoll, 2, -1) != -1) {
-        for (i = 0; i < 2; i++) {
-            if (epoll[i].data.fd != udev_monitor->pfd[0]) {
-                continue;
-            }
+    poll_fd.fd = udev_monitor->ifd;
+    poll_fd.events = POLLIN;
 
-            read(udev_monitor->pfd[0], data, 1);
-            write(udev_monitor->pfd[1], "1", 1);
-            pthread_barrier_wait(&udev_monitor->barrier);
-            return NULL;
+    while (1) {
+        if (poll(&poll_fd, 1, -1) == -1) {
+            continue;
+        }
+        // exit on fd close
+        if (!(poll_fd.revents & POLLIN)) {
+            break;
```

Comment:
We can't rely on this. I suggest you to keep self-pipe trick.

```c
// assume that fds[1] is self-pipe fd
if (fds[1].revents & POLLIN) {
    break;
}
```

`poll(2)`

```
For a discussion of what may happen if a file descriptor being
monitored by poll() is closed in another thread, see select(2).
```

`select(2)`

```
Multithreaded applications
       If a file descriptor being monitored by select() is closed in another
       thread, the result is unspecified.  On some UNIX systems, select()
       unblocks and returns, with an indication that the file descriptor is
       ready (a subsequent I/O operation will likely fail with an error,
       unless another process reopens file descriptor between the time
       select() returned and the I/O operation is performed).  On Linux (and
       some other systems), closing the file descriptor in another thread has
       no effect on select().  In summary, any application that relies on a
       particular behavior in this scenario must be considered buggy.
```

Reply:
Yeah I saw this right after I pushed. Made killing X take a really long time. I fixed this in the last push to use an event fd. I don't have anything against using a pipe though